src/compiler/checker.ts

```diff
@@ -4092,7 +4106,15 @@ namespace ts {
                         const resolvedSymbol = resolveName(param, paramSymbol.name, SymbolFlags.Value, undefined, undefined);
                         paramSymbol = resolvedSymbol;
                     }
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {

```

Comment:
Actually, since we don't care about JSConstructSignature case, you should just say `if (i === 0 && paramSymbol.name === "this")`.